// @tanstack/router-generator/dist/esm/index.js
import { baseConfigSchema, configSchema, getConfig, resolveConfigPath } from "./config.js";
import { Generator } from "./generator.js";
import { capitalize, checkRouteFullPathUniqueness, cleanPath, determineInitialRoutePath, format, inferFullPath, multiSortBy, removeExt, removeLeadingSlash, removeTrailingSlash, removeUnderscores, replaceBackslash, resetRegex, routePathToVariable, trimPathLeft, writeIfDifferent } from "./utils.js";
import { getRouteNodes } from "./filesystem/physical/getRouteNodes.js";
import { getRouteNodes as getRouteNodes2 } from "./filesystem/virtual/getRouteNodes.js";
import { rootPathId } from "./filesystem/physical/rootPathId.js";
import { ensureStringArgument } from "./transform/utils.js";
export {
  Generator,
  baseConfigSchema,
  capitalize,
  checkRouteFullPathUniqueness,
  cleanPath,
  configSchema,
  determineInitialRoutePath,
  ensureStringArgument,
  format,
  getConfig,
  inferFullPath,
  multiSortBy,
  getRouteNodes as physicalGetRouteNodes,
  removeExt,
  removeLeadingSlash,
  removeTrailingSlash,
  removeUnderscores,
  replaceBackslash,
  resetRegex,
  resolveConfigPath,
  rootPathId,
  routePathToVariable,
  trimPathLeft,
  getRouteNodes2 as virtualGetRouteNodes,
  writeIfDifferent
};
