// @tanstack/router-generator/dist/cjs/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const config = require("./config.cjs");
const generator = require("./generator.cjs");
const utils = require("./utils.cjs");
const getRouteNodes = require("./filesystem/physical/getRouteNodes.cjs");
const getRouteNodes$1 = require("./filesystem/virtual/getRouteNodes.cjs");
const rootPathId = require("./filesystem/physical/rootPathId.cjs");
const utils$1 = require("./transform/utils.cjs");
exports.baseConfigSchema = config.baseConfigSchema;
exports.configSchema = config.configSchema;
exports.getConfig = config.getConfig;
exports.resolveConfigPath = config.resolveConfigPath;
exports.Generator = generator.Generator;
exports.capitalize = utils.capitalize;
exports.checkRouteFullPathUniqueness = utils.checkRouteFullPathUniqueness;
exports.cleanPath = utils.cleanPath;
exports.determineInitialRoutePath = utils.determineInitialRoutePath;
exports.format = utils.format;
exports.inferFullPath = utils.inferFullPath;
exports.multiSortBy = utils.multiSortBy;
exports.removeExt = utils.removeExt;
exports.removeLeadingSlash = utils.removeLeadingSlash;
exports.removeTrailingSlash = utils.removeTrailingSlash;
exports.removeUnderscores = utils.removeUnderscores;
exports.replaceBackslash = utils.replaceBackslash;
exports.resetRegex = utils.resetRegex;
exports.routePathToVariable = utils.routePathToVariable;
exports.trimPathLeft = utils.trimPathLeft;
exports.writeIfDifferent = utils.writeIfDifferent;
exports.physicalGetRouteNodes = getRouteNodes.getRouteNodes;
exports.virtualGetRouteNodes = getRouteNodes$1.getRouteNodes;
exports.rootPathId = rootPathId.rootPathId;
exports.ensureStringArgument = utils$1.ensureStringArgument;
