// eslint-plugin-better-tailwindcss/lib/options/callees/tv.js
import { MatcherType } from "../../types/rule.js";
export const TV_STRINGS = [
    "tv",
    [
        {
            match: MatcherType.String
        }
    ]
];
export const TV_VARIANT_VALUES = [
    "tv",
    [
        {
            match: MatcherType.ObjectValue,
            pathPattern: "^variants.*$"
        }
    ]
];
export const TV_BASE_VALUES = [
    "tv",
    [
        {
            match: MatcherType.ObjectValue,
            pathPattern: "^base$"
        }
    ]
];
export const TV_SLOTS_VALUES = [
    "tv",
    [
        {
            match: MatcherType.ObjectValue,
            pathPattern: "^slots.*$"
        }
    ]
];
export const TV_COMPOUND_VARIANTS_CLASS = [
    "tv",
    [
        {
            match: MatcherType.ObjectValue,
            pathPattern: "^compoundVariants\\[\\d+\\]\\.(?:className|class).*$"
        }
    ]
];
export const TV_COMPOUND_SLOTS_CLASS = [
    "tv",
    [
        {
            match: MatcherType.ObjectValue,
            pathPattern: "^compoundSlots\\[\\d+\\]\\.(?:className|class).*$"
        }
    ]
];
/** @see https://github.com/nextui-org/tailwind-variants?tab=readme-ov-file */
export const TV = [
    TV_STRINGS,
    TV_VARIANT_VALUES,
    TV_COMPOUND_VARIANTS_CLASS,
    TV_BASE_VALUES,
    TV_SLOTS_VALUES,
    TV_COMPOUND_SLOTS_CLASS
];
